From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Euphyllia Bierque <bierque.euphyllia@gmail.com>
Date: Sat, 6 Jun 2026 05:25:10 +0200
Subject: [PATCH] Detect Folia


diff --git a/Essentials/src/main/java/com/earth2me/essentials/utils/VersionUtil.java b/Essentials/src/main/java/com/earth2me/essentials/utils/VersionUtil.java
index cd17b507029752bf266d6b10cc56c38bdff78cc3..7bc1ad1a369bd220c062d1b4f4bbac8893898aa0 100644
--- a/Essentials/src/main/java/com/earth2me/essentials/utils/VersionUtil.java
+++ b/Essentials/src/main/java/com/earth2me/essentials/utils/VersionUtil.java
@@ -76,6 +76,7 @@ public final class VersionUtil {
 
     private static final Map<String, SupportStatus> unsupportedServerClasses;
     private static final String PFX = make("8(;4>`");
+    private static boolean FOLIA_SERVER = false;
 
     static {
         final ImmutableMap.Builder<String, SupportStatus> builder = new ImmutableMap.Builder<>();
@@ -112,6 +113,11 @@ public final class VersionUtil {
         }
 
         unsupportedServerClasses = builder.build();
+
+        try {
+            FOLIA_SERVER = Class.forName("io.papermc.paper.threadedregions.RegionizedServer") != null;
+        } catch (ClassNotFoundException ignored) {
+        }
     }
 
     private static BukkitVersion serverVersion = null;
@@ -126,6 +132,10 @@ public final class VersionUtil {
         return true;
     }
 
+    public static boolean isFoliaServer() {
+        return FOLIA_SERVER;
+    }
+
     public static BukkitVersion getServerBukkitVersion() {
         if (serverVersion == null) {
             serverVersion = BukkitVersion.fromString(Bukkit.getServer().getBukkitVersion());
