From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Euphyllia Bierque <bierque.euphyllia@gmail.com>
Date: Sat, 6 Jun 2026 06:53:14 +0200
Subject: [PATCH] Fix commands command in afk


diff --git a/Essentials/src/main/java/com/earth2me/essentials/User.java b/Essentials/src/main/java/com/earth2me/essentials/User.java
index 33b973544f689c0b1d524e3611da163ed71a6378..38a18068e5f5f06ac582fd6686a1ba101632f02a 100644
--- a/Essentials/src/main/java/com/earth2me/essentials/User.java
+++ b/Essentials/src/main/java/com/earth2me/essentials/User.java
@@ -910,6 +910,7 @@ public class User extends UserData implements Comparable<User>, IMessageRecipien
                 }
             } else {
                 // If `afk-timeout-commands` in config.yml is populated, execute the command(s) instead of kicking the player.
+                ess.getGlobalRegionScheduler().execute(ess, () -> {
                 for (final String command : ess.getSettings().getAfkTimeoutCommands()) {
                     if (command == null || command.isEmpty()){
                         continue;
@@ -917,7 +918,7 @@ public class User extends UserData implements Comparable<User>, IMessageRecipien
                     // Replace placeholders in the command with actual values.
                     final String cmd = command.replace("{USERNAME}", getName()).replace("{KICKTIME}", String.valueOf(kickTime));
                     ess.getServer().dispatchCommand(ess.getServer().getConsoleSender(), cmd);
-                }
+                }});
             }
         }
         final long autoafk = ess.getSettings().getAutoAfk();
