From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Euphyllia Bierque <bierque.euphyllia@gmail.com>
Date: Sat, 6 Jun 2026 06:54:31 +0200
Subject: [PATCH] Fix for crash if a player leaves while someone else /invsee
 them


diff --git a/Essentials/src/main/java/com/earth2me/essentials/EssentialsPlayerListener.java b/Essentials/src/main/java/com/earth2me/essentials/EssentialsPlayerListener.java
index 4f17ad65eb2a49f065e33b998e12c33db54287bf..511871671e75bb1edfc22e4b26f98f1df0d9eaf4 100644
--- a/Essentials/src/main/java/com/earth2me/essentials/EssentialsPlayerListener.java
+++ b/Essentials/src/main/java/com/earth2me/essentials/EssentialsPlayerListener.java
@@ -308,13 +308,13 @@ public class EssentialsPlayerListener implements Listener {
         }
 
         final ArrayList<HumanEntity> viewers = new ArrayList<>(user.getBase().getInventory().getViewers());
-        for (final HumanEntity viewer : viewers) {
+        for (final HumanEntity viewer : viewers) {  ess.getEntityScheduler(viewer).run(ess, scheduledTask -> {
             if (viewer instanceof Player) {
                 final User uviewer = ess.getUser((Player) viewer);
                 if (uviewer.isInvSee()) {
                     uviewer.getBase().closeInventory();
                 }
-            }
+            }}, null);
         }
 
         user.updateActivity(false, AfkStatusChangeEvent.Cause.QUIT);
